pull request for demo

soborok opened the pull request:
## Heroku

https://to-catch-a-troll.herokuapp.com
## FEATURES
1. User signup/login
2. Scrape a webpage(s)
3. Search keyword in scraped webpages
## Readiness for Demos
1. deployed to heroku: user model working, solr tasks yet implemented
2. added/set up links to heroku, github, trello, travis on [wiki](https://github.com/bison-2014/phase-3-guide/wiki/Groups#tocatchatroll)
3. set up travis, simplecov
## Screenshots

![1_login_form](https://cloud.githubusercontent.com/assets/8992722/6212663/7cb8c908-b5ad-11e4-9ea7-57266cf0b81e.png)
![2_signup_empty](https://cloud.githubusercontent.com/assets/8992722/6212674/942654b6-b5ad-11e4-9540-703acc654f73.png)
![3_singedin_success](https://cloud.githubusercontent.com/assets/8992722/6212675/9429d938-b5ad-11e4-8312-967303cdaca5.png)
![4_search_new](https://cloud.githubusercontent.com/assets/8992722/6212677/942f3e8c-b5ad-11e4-810a-3118d9780d57.png)
![5_search_results1](https://cloud.githubusercontent.com/assets/8992722/6212676/942f2c94-b5ad-11e4-8b03-cda359774ef4.png)
![5_search_results2](https://cloud.githubusercontent.com/assets/8992722/6212678/9432a5fe-b5ad-11e4-94d6-659f0a82a1d1.png)
![6_scraper_list_of_scrapedwebsits](https://cloud.githubusercontent.com/assets/8992722/6212679/9437e686-b5ad-11e4-9171-e4309a3ff37b.png)
![7_scraper_scrape_new](https://cloud.githubusercontent.com/assets/8992722/6212680/943a556a-b5ad-11e4-9dd3-192a8296d3dd.png)
![8_signedout_success](https://cloud.githubusercontent.com/assets/8992722/6212728/01da1880-b5ae-11e4-9509-1c85efb640e2.png)

soborok committed:
resolved spec file name issue

dalrayes committed:
Merge pull request #42 from soborok/layouts-nav

Layouts nav

soborok committed:
Merge pull request #43 from dalrayes/master

Updated search results view page - working again

soborok committed:
changed depth, edit nav link, added resque route

dalrayes committed:
added tests for Page controller

alcastaneda committed:
Merge pull request #44 from dalrayes/master

added tests for Page controller

dalrayes committed:
added normalize.css

soborok committed:
Merge pull request #45 from dalrayes/master

Added normalize.css

soborok committed:
moved simplecov code in spec/ helper files

pgblu committed:
Merge pull request #46 from soborok/few-changes

testing Simplecov

pgblu committed:
custom_crawler spec first commit

pgblu committed:
custom_crawler spec first commit

pgblu committed:
custom_crawler spec first commit

pgblu committed:
custom_crawler spec first commit

pgblu committed:
custom_crawler spec first commit

pgblu committed:
custom_crawler spec first commit

soborok committed:
added scraper_controller test code

dalrayes committed:
Merge pull request #47 from soborok/test-scraper

added scraper_controller test code

pgblu committed:
Merge branch 'master' of https://github.com/pgblu/to-catch-a-troll

alcastaneda committed:
adding upstream master changes

alcastaneda committed:
adding upstream master changes

pgblu committed:
craigslisthelp

pgblu committed:
Merge branch 'master' of https://github.com/pgblu/to-catch-a-troll

pgblu committed:
Merge branch 'zipdata'

pgblu committed:
commented out stub of image filter

dalrayes committed:
Updated search result display: Highlights working

pgblu committed:
added craigslist csv file

pgblu committed:
disabled custom crawler feature tests for now

soborok committed:
Merge pull request #48 from dalrayes/master

Updated search result display: Highlights working

soborok committed:
added search_query model/controller/views/routes/migration

soborok committed:
added search_query model/controller/views/routes/migration

pgblu committed:
also disable model tests for custom crawler

dalrayes committed:
Merge pull request #51 from pgblu/master

disabled crawler model spec

pgblu committed:
craigslist lookup functionality, to be encapsulated for: sanitizer options, user search range delimiting, and recursive get logic only

dalrayes committed:
Updated results page with formatting

dalrayes committed:
Updated results page with formatting

alcastaneda committed:
Merge pull request #52 from soborok/search-query

added search_query model/controller/views/routes/migration

dalrayes committed:
updated the results to take in long strings of info

dalrayes committed:
updated the results to take in long strings of info

alcastaneda committed:
made search query edit view form and controller

alcastaneda committed:
made search query edit view form and controller

pgblu committed:
models changes to incorporate target table logic in custom_crawler & user

pgblu committed:
models changes to incorporate target table logic in custom_crawler & user

pgblu committed:
sanitizing and injection of target object logic into custom crawler and task worker classes

pgblu committed:
sanitizing and injection of target object logic into custom crawler and task worker classes

pgblu committed:
sanitizing and injection of target object logic into custom crawler and task worker classes

pgblu committed:
targets controller begun; target and craigslist-lookup models, and migration for targets

pgblu committed:
targets controller begun; target and craigslist-lookup models, and migration for targets

pgblu committed:
targets controller begun; target and craigslist-lookup models, and migration for targets

pgblu committed:
targets controller begun; target and craigslist-lookup models, and migration for targets

dalrayes committed:
search integrated with user search query

dalrayes committed:
search integrated with user search query

pgblu committed:
incorporating changes from upstream

pgblu committed:
incorporating changes from upstream

pgblu committed:
resolved conflicts

rmw commented on README.md:
```diff
@@ -1,2 +1,21 @@
 # ToCatchATroll
-a resource for monitoring negative references to and vulnerabilities in one's personal data
+**Group Members & Github Handles**
```
:+1: Awesomesauce!! Thank you!

rmw commented on app/controllers/pages_controller.rb:
```diff
@@ -0,0 +1,22 @@
+class PagesController < ApplicationController
+
+# before_action :authenticate_user!
```
No commented out code in master! :smiling_imp:

rmw commented on app/controllers/pages_controller.rb:
```diff
@@ -0,0 +1,22 @@
+
+  def search
+    if params[:search].blank?
+      render 'search'
```
Does this work?  You may need to say `return render 'search'`

rmw commented on app/controllers/pages_controller.rb:
```diff
@@ -0,0 +1,22 @@
+      render 'search'
+    end
+
+    if Page.count > 0
```
It feels like the `Page` model should be responsible for handling this searching.  You can create a class method on `Page`:

``` ruby
class Page
  def self.term_search(terms)
    return nil if count <= 0

    result = search do
      fulltext(terms) do
        highlight :body
      end
    end

    return result.hits
  end
end
```

Then your controller method might look like:

``` ruby
def search
  return render 'search' if params[:search].blank?

  @results = Page.term_search(params[:search])

  if @results.nil?
    @message = "Start your first search"
  end
end
```

rmw commented on app/controllers/scrapers_controller.rb:
```diff
@@ -0,0 +1,29 @@
+class ScrapersController < ApplicationController
```
This is really cool that you can configure scrapers like this!  Is this something that any user can do or only an admin?  If only admins, let's talk tomorrow about restricting certain users access.

rmw commented on app/controllers/search_queries_controller.rb:
```diff
@@ -0,0 +1,31 @@
+  def create
+    query = SearchQuery.create(user: current_user)
+    query.update_attributes(search_query_params)
+    query.save
```
I'd put a new line here.  A good rule of thumb is never have more than 3 lines in a row and group things as you would a paragraph.

rmw commented on app/controllers/users/confirmations_controller.rb:
```diff
@@ -0,0 +1,28 @@
+class User::ConfirmationsController < Devise::ConfirmationsController
+  # GET /resource/confirmation/new
```
Unless you are going to customize this controller, you don't need to have  the file at all.

rmw commented on app/models/custom_crawler.rb:
```diff
@@ -0,0 +1,25 @@
+class CustomCrawler
```
:+1: service object!!

rmw commented on app/models/custom_crawler.rb:
```diff
@@ -0,0 +1,25 @@
+  def recursive_get(base_url, depth = 2)
+    unless depth < 0
+      begin
+        f = @cw.get(base_url)
```
Can we name f better?

rmw commented on app/models/custom_crawler.rb:
```diff
@@ -0,0 +1,25 @@
+      begin
+        f = @cw.get(base_url)
+      rescue
+        puts "a connection was refused, move on"
```
Instead of using `puts` here, you should use the [Rails logger] with the appropriate log level.  For example, `Rails.logger.info "Connection refused, moving on"`.  Also, generally errors are written in present tense. :memo:

rmw commented on app/models/search_query.rb:
```diff
@@ -0,0 +1,9 @@
+	validates :user, presence: true
+
+	def search_query
+		"#{self.first_name} #{self.last_name} #{self.address} #{self.zipcode} #{self.phone_number}"
```
What's with the weird indentation here? Make sure everyone has Sublime set to use spaces instead of tabs and 2 spaces for a tab.

rmw replied:
Also, you don't need to use `self` for readers.

rmw commented on app/views/pages/_results.html.erb:
```diff
@@ -0,0 +1,16 @@
+<h1>Search Results</h1>
+<h2>Results found:<%= @results_found.count %></h2>
+
+<% @usersearch.hits.each do |hit| %>
```
You could use a [partial with collection here](http://guides.rubyonrails.org/layouts_and_rendering.html#rendering-collections).

rmw commented on bin/rails:
```diff
@@ -6,3 +6,9 @@ end
 require 'rails/commands'
+
+if ENV['RAILS_ENV'] == 'test'
+  require 'simplecov'
```
You should be able to put this code in your `rails_helper.rb`.

rmw commented on bin/rails:
```diff
@@ -6,3 +6,9 @@ end
+if ENV['RAILS_ENV'] == 'test'
+  require 'simplecov'
+  SimpleCov.start 'rails'
+  puts "required simplecov"
```
Don't use `puts` within your Rails application.  If you want to get info from your app, use `Rails.logger` (see comment above.)

rmw commented on config/routes.rb:
```diff
@@ -1,10 +1,32 @@
+    end
+
+    unauthenticated do
+      root 'devise/sessions#new', as: :unauthenticated_root
```
:+1: Nice use of named routes.

rmw commented on config/routes.rb:
```diff
@@ -1,10 +1,32 @@
+
+  resources :scrapers, only: [:index, :new, :create]
+  get '/search' => 'pages#search'
+  resources :pages, :only => [:index, :new, :show, :create, :update, :delete]
```
Hmm ... why do you not have an `edit` route, but you do have an `update` route?

rmw commented on spec/controllers/pages_controller_spec.rb:
```diff
@@ -0,0 +1,12 @@
+  describe 'GET #search' do
+    it 'renders the search page' do
+      get :search
+      expect(response.status).to eq(200)
```
Make sure your test and the message are in line.  In this case the test is _actually_ testing if the page returns successful.  To test that it renders the search page we'd say something like `expect(response).to render_template(:search)`.

rmw commented on spec/controllers/scrapers_controller_spec.rb:
```diff
@@ -0,0 +1,46 @@
+RSpec.describe ScrapersController, type: :controller do
+
+	describe "GET #index" do
+		xit "lists all scrapers" do
```
This indentation is crazy town.

rmw commented on spec/controllers/scrapers_controller_spec.rb:
```diff
@@ -0,0 +1,46 @@
+		xit "lists all scrapers" do
+			scraper = Scraper.create
+      get :index
+      expect(assigns(:scrapers)).to eq([scrapers])
```
I think you just need to say `eq([scraper])` (without the `s`) to fix this test.

rmw commented on spec/controllers/scrapers_controller_spec.rb:
```diff
@@ -0,0 +1,46 @@
+      expect(response).to render_template("new")
+		end
+
+		xit "renders the actual template" do
```
This test is probably unnecessary, but if you wanted to do something like this, you need to include `render_views` in your test.

rmw commented on spec/features/custom_crawler_spec.rb:
```diff
@@ -0,0 +1,22 @@
+    let (:page_count) { Page.all.count }
+
+    it "increases the number of page objects by 1" do
+      ccrawler.get(url, 0)
```
make sure you are not actually making web requests here.  Take a look at the [FakeWeb gem](https://github.com/chrisk/fakeweb) to mock out the http requests.  You may also want to use Fake to generate some random urls.

rmw replied:
Oh wait, this test is not a feature test and looks like its a duplicate of the model test.  Remove this one.

rmw commented:
Dear @soborok @pgblu @dalrayes @alcastaneda 
cc @m3rabb @dukegreene 

Pulling all these new technologies together and deploying them to production is huge feat! :confetti_ball: :congratulations:  Overall, your Rails app is a solid foundation for adding more UI features and styling. 

Also, thank you for being diligent and following all the directions! :tada: 

I think there are two areas that could use improvement:

#### Indentation

There were a couple files with crazy indentation.  Make sure everyone has the same tab/space settings in Sublime.  It's really hard to read code that isn't consistently indented.

#### Testing

You have a really good start on the testing front, but there's still a ways to go.

Make sure your controller tests test success and failure, especially for post requests.  

Devise very kindly gives use [built in Rspec helpers](https://github.com/plataformatec/devise#test-helpers).  There are a number of ways to test Resque and [this gem](https://github.com/leshill/resque_spec#resque-with-specs) may be worth looking into.

You'll want to look at something like [FakeWeb](https://github.com/chrisk/fakeweb) for testing the CustomCrawler.  It would be really good for employers to see solid testing around this.

pgblu commented:
Thanks for these links! We look forward to your further feedback.

On Mon, Feb 16, 2015 at 6:09 PM, Rebecca Miller-Webster <
notifications@github.com> wrote:

> Dear @soborok https://github.com/soborok @pgblu
> https://github.com/pgblu @dalrayes https://github.com/dalrayes
> @alcastaneda https://github.com/alcastaneda
> cc @m3rabb https://github.com/m3rabb @dukegreene
> https://github.com/dukegreene
> 
> Pulling all these new technologies together and deploying them to
> production is huge feat! [image: :confetti_ball:] [image:
> :congratulations:] Overall, your Rails app is a solid foundation for
> adding more UI features and styling.
> 
> Also, thank you for being diligent and following all the directions! [image:
> :tada:]
> 
> I think there are two areas that could use improvement:
> Indentation
> 
> There were a couple files with crazy indentation. Make sure everyone has
> the same tab/space settings in Sublime. It's really hard to read code that
> isn't consistently indented.
> Testing
> 
> You have a really good start on the testing front, but there's still a
> ways to go.
> 
> Make sure your controller tests test success and failure, especially for
> post requests.
> 
> Devise very kindly gives use built in Rspec helpers
> https://github.com/plataformatec/devise#test-helpers. There are a
> number of ways to test Resque and this gem
> https://github.com/leshill/resque_spec#resque-with-specs may be worth
> looking into.
> 
> You'll want to look at something like FakeWeb
> https://github.com/chrisk/fakeweb for testing the CustomCrawler. It
> would be really good for employers to see solid testing around this.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/rmw/to-catch-a-troll/pull/1#issuecomment-74594509.

## 

Philipp Blume

alcastaneda committed:
added edit search query view and controller

alcastaneda committed:
added edit search query view and controller

dalrayes committed:
Merge pull request #54 from pgblu/master

Changes in a nutshell: seeds, migrations, models, and stub of controller to support target and craigslist_lookup functionality

alcastaneda committed:
commented out sessions controller test, not working

alcastaneda committed:
commented out sessions controller test, not working

alcastaneda committed:
deleted sessions controller test, not working

alcastaneda committed:
deleted sessions controller test, not working

soborok committed:
Merge pull request #57 from alcastaneda/master

made search query edit view form and controller

dalrayes committed:
Added access to user query to page controller

dalrayes committed:
Added access to user query to page controller

dalrayes committed:
Updated search_strings method

dalrayes committed:
Updated search_strings method

dalrayes committed:
updated user landing page to have profile and recent finds

dalrayes committed:
updated user landing page to have profile and recent finds

dalrayes committed:
Updated home route

dalrayes committed:
Updated home route

dalrayes committed:
added user query edit link to landing page

dalrayes committed:
added user query edit link to landing page

dalrayes committed:
modified search_query route for new routes, edit nav bar for more actual ux

dalrayes committed:
modified search_query route for new routes, edit nav bar for more actual ux

soborok committed:
changed depth to 0 for websolr test

soborok committed:
added redis addon to heroku

soborok committed:
added Resque.redis = REDIS to resque_auth.rb

soborok committed:
added solr production path, hostname

soborok committed:
added solr production path, hostname

soborok committed:
changed redis setting in redis.rb

soborok committed:
set heroku config REDIS_PROVIDER

dalrayes committed:
Working on search function for user profile

dalrayes committed:
Working on search function for user profile

alcastaneda committed:
Merge pull request #59 from dalrayes/master

Updated the user landing page and options - fixed for Travis error

alcastaneda committed:
added target field and associations to pages

alcastaneda committed:
added target field and associations to pages

alcastaneda committed:
added target field and associations to pages

alcastaneda committed:
added target field and associations to pages

alcastaneda committed:
added new column to pages

alcastaneda committed:
added new column to pages

dalrayes committed:
Merge pull request #60 from alcastaneda/master

add target field to pages

soborok committed:
adding procfile to manage assets on heroku during deploy

soborok committed:
adding sidekiq to Procfile

soborok committed:
configure web worker to use bundler

soborok committed:
update the worker to use the correct rake task

soborok committed:
rebased with heroku websolr, redis, reque setup

pgblu committed:
crawler has target

soborok committed:
resolving merge conflict of Procfile

alcastaneda committed:
add clockwork job for scraper

alcastaneda committed:
add clockwork job for scraper

soborok committed:
Merge pull request #62 from alcastaneda/master

add clockwork job for scraper

pgblu committed:
resolved conflict

pgblu committed:
merge conflict resolved in app.models/target.rb

pgblu committed:
git hell

pgblu committed:
git hell

pgblu committed:
git hell

pgblu committed:
git hell

pgblu committed:
custom_crawler model clean

pgblu committed:
custom_crawler model clean

pgblu committed:
merge

pgblu committed:
merge

pgblu committed:
cleaning up conflicts

pgblu committed:
cleaning up conflicts

soborok committed:
reseting solr to resove http error

soborok committed:
added error message to rescue

soborok committed:
verify redis init is running

soborok committed:
more debugging

dalrayes committed:
updated filtering for search - needs to be linked to actual target objects

dalrayes committed:
updated filtering for search - needs to be linked to actual target objects

soborok committed:
using the default config from redisttogo

soborok committed:
Merge pull request #65 from dalrayes/master

updated filtering for search - needs to be linked to actual target obj..

pgblu committed:
fixed Sanitize, now saving pages

pgblu committed:
resolved merge conflict

pgblu committed:
resolving merge conflict

pgblu committed:
neutered the test

soborok committed:
Merge pull request #66 from pgblu/master

Sanitize is working, Page is being saved

soborok committed:
added heroku setting for redis, websolr

dalrayes committed:
Merge pull request #68 from soborok/heroku-setting

added heroku setting for redis, websolr

soborok committed:
resolving merge conflict

soborok committed:
temp

soborok committed:
Merge branch 'master' into seo

soborok committed:
removed debugging text

pgblu commented:
Dear Rebecca,

I'm a bit at a loss how to test my CustomCrawler. I saw your FakeWeb link,
and I understand it without much trouble, but how do I use it to test what
are essentially 'side-effects' that write to the database? It seems that
the Page objects that get written are not exposed to the test suite (?)

Here is the code for your reference:

class CustomCrawler

  def initialize(target_id )
    option_hash = { crawl_limit_by_page: true }
    @cw = Cobweb.new(option_hash)
    @target = Target.find_by(id: target_id)
  end

  def recursive_get(base_url, depth = 2)
    unless depth < 0
      if @target.sanitize_options
       options = eval(@target.sanitize_options)
      else
        options = { elements: ['html', 'div', 'p', 'span'] }
      end

```
  begin
    f = @cw.get(base_url)
  rescue => e
    puts "a connection is refused, move on: ERROR #{e.inspect}"
  end
  if page = Page.find_by(base_url: base_url)
    page.destroy
  end
  if f && f[:status_code] == 200 && ![".jpg", ".png", ".gif", ".tiff",
```

".swf"].any? {|extension| f[:base_url].include? (extension)}
        f[:body].force_encoding('iso-8859-1').encode('utf-8')
        sanitized_file = Sanitize.document(f[:body], options)
        Page.create(base_url: base_url, body: sanitized_file, target_id: @
target.id)
        f[:links][:links].each { |link| recursive_get(link, depth-1) }
      end
    end
    nil
  end

end

pgblu commented:
And a link:
https://github.com/bison-2014/to-catch-a-troll/blob/master/app/models/custom_crawler.rb

On Tue, Feb 17, 2015 at 8:18 PM, Philipp Blume pgbluster@gmail.com wrote:

> Dear Rebecca,
> 
> I'm a bit at a loss how to test my CustomCrawler. I saw your FakeWeb link,
> and I understand it without much trouble, but how do I use it to test what
> are essentially 'side-effects' that write to the database? It seems that
> the Page objects that get written are not exposed to the test suite (?)
> 
> Here is the code for your reference:
> 
> class CustomCrawler
> 
>   def initialize(target_id )
>     option_hash = { crawl_limit_by_page: true }
>     @cw = Cobweb.new(option_hash)
>     @target = Target.find_by(id: target_id)
>   end
> 
>   def recursive_get(base_url, depth = 2)
>     unless depth < 0
>       if @target.sanitize_options
>        options = eval(@target.sanitize_options)
>       else
>         options = { elements: ['html', 'div', 'p', 'span'] }
>       end
> 
> ```
>   begin
>     f = @cw.get(base_url)
>   rescue => e
>     puts "a connection is refused, move on: ERROR #{e.inspect}"
>   end
>   if page = Page.find_by(base_url: base_url)
>     page.destroy
>   end
>   if f && f[:status_code] == 200 && ![".jpg", ".png", ".gif", ".tiff",
> ```
> 
> ".swf"].any? {|extension| f[:base_url].include? (extension)}
>         f[:body].force_encoding('iso-8859-1').encode('utf-8')
>         sanitized_file = Sanitize.document(f[:body], options)
>         Page.create(base_url: base_url, body: sanitized_file, target_id: @
> target.id)
>         f[:links][:links].each { |link| recursive_get(link, depth-1) }
>       end
>     end
>     nil
>   end
> 
> end

## 

Philipp Blume

dalrayes committed:
updated for conflict

dalrayes committed:
Merge pull request #69 from soborok/scraper-working-heroku

Scraper working heroku

soborok committed:
Merge pull request #70 from dalrayes/master

Updated CSS with Foundation

pgblu committed:
custom crawler testing

alcastaneda committed:
add blank mailer templates

alcastaneda committed:
added mailer config

alcastaneda committed:
edit mailer config

alcastaneda committed:
added usermaler model and migration

dalrayes committed:
removed back button

alcastaneda committed:
add mailer controller template

alcastaneda committed:
add email message and button to mailer index view for testing

rmw commented:
@pgblu - what do you mean they are side effects? 

FakeWeb will block all http requests and then you can specify the http response for that request (status and html usually).

Then you test your code and methods just like any other.  Off the top of my head:

```
describe CustomCrawler do
  describe "#initialize" do
    it "finds the target" do
       target = FactoryGirl.create(:target)
      allow(Target).to receive(:find_by).with(id: target.id).and_return(target)
      CustomCrawl.new(target.id)
    end

    # etc
  end

  describe "#recursive_get" do
    let(:crawler) { CustomCrawler.new(1) }
    # here you would specify the fake web responses to your get requests - assume you specify 1 link with your html
    it "creates 2 page objects" do
      expect {
         crawler.recursive_get('example.com')
       }.to change(Page, :count).by(2)
    end
  end
end
```

You are overthinking this.  It's not magic.  You wrote this code and you _know_ what you expect it to do.  You just need to put it into rspec syntax.  Write out what you expect each line to do (in your own words) and then write a test for that.

alcastaneda committed:
add controller code for mailer testing

pgblu commented:
By side effects I mean the function doesn't return anything but simply
modifies the database. I will strive to underthink it now. :-)

On Tue, Feb 17, 2015 at 10:44 PM, Rebecca Miller-Webster <
notifications@github.com> wrote:

> @pgblu https://github.com/pgblu - what do you mean they are side
> effects?
> 
> FakeWeb will block all http requests and then you can specify the http
> response for that request (status and html usually).
> 
> Then you test your code and methods just like any other. Off the top of my
> head:
> 
> describe CustomCrawler do
>   describe "#initialize" do
>     it "finds the target" do
>        target = FactoryGirl.create(:target)
>       allow(Target).to receive(:find_by).with(id: target.id).and_return(target)
>       CustomCrawl.new(target.id)
>     end
> 
> ```
> # etc
> ```
> 
>   end
> 
>   describe "#recursive_get" do
>     let(:crawler) { CustomCrawler.new(1) }
>     # here you would specify the fake web responses to your get requests - assume you specify 1 link with your html
>     it "creates 2 page objects" do
>       expect {
>          crawler.recursive_get('example.com')
>        }.to change(Page, :count).by(2)
>     end
>   end
> end
> 
> You are overthinking this. It's not magic. You wrote this code and you
> _know_ what you expect it to do. You just need to put it into rspec
> syntax. Write out what you expect each line to do (in your own words) and
> then write a test for that.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/rmw/to-catch-a-troll/pull/1#issuecomment-74811310.

## 

Philipp Blume

dalrayes committed:
Target filters working

alcastaneda committed:
added email test controller and view

dalrayes committed:
updated filter

dalrayes committed:
updated page title

soborok committed:
Merge pull request #71 from dalrayes/master

Updated the page filter with the 3 sites

soborok committed:
added target_id to scrapers table

soborok committed:
changed to use target id when creating scraper and applied to view files from typing in to selecting

soborok committed:
fixed spcaing

soborok committed:
commented out scraper controller test code

soborok committed:
fixed typo

soborok committed:
commented out scraper controller test code

dalrayes committed:
Merge pull request #72 from soborok/target-selectbox

select target url to scrape

pgblu committed:
extra line feeds removed from crawler model

alcastaneda committed:
finish test email feature

pgblu committed:
ajaxifying form

alcastaneda committed:
working mailer

alcastaneda committed:
Merge pull request #73 from pgblu/master

ajax on search query creation

alcastaneda committed:
customized email messagegit add .git add .!

soborok committed:
Merge pull request #76 from alcastaneda/master

Notification Mailer

dalrayes committed:
Added flags model and fully functional

soborok committed:
resolved merge conflict

dalrayes committed:
Updated flag to sort by most recent, adding flag icons files

alcastaneda committed:
Merge pull request #77 from soborok/page-targets

changed to checkbox from radio button

pgblu committed:
form page ajax

dalrayes committed:
Updated schema with db:migrate

dalrayes committed:
Merge pull request #80 from dalrayes/master

Added flags model and migration

soborok committed:
Merge pull request #81 from pgblu/master

form page ajax

dalrayes committed:
updated header with flags

dalrayes committed:
updated pages controller for 'chosen_sites'

dalrayes committed:
Merge pull request #82 from dalrayes/master

Updated variable in pages controller (chosen_sites)

pgblu committed:
crawler code perfect now

pgblu committed:
Merge branch 'custombetter'

pgblu committed:
Merge pull request #83 from pgblu/master

Custom Crawler is now as good as it needs to be

soborok committed:
fixed sunspot:solr connection issue

dalrayes committed:
Merge pull request #84 from soborok/solr-connection-issue

fixed sunspot:solr connection issue

pgblu committed:
some view improvements

soborok committed:
resolved merge conflict

soborok committed:
added cron job

pgblu committed:
Merge pull request #85 from pgblu/master

some view improvements

dalrayes committed:
CSS changes to login form buttons

soborok committed:
figuring out website checkbox

soborok committed:
adding website

dalrayes committed:
Updated readme description

dalrayes committed:
Merge pull request #86 from alcastaneda/refactored_search

Refactored search

soborok committed:
Merge pull request #87 from dalrayes/master

Updated README

soborok committed:
added screenshot

pgblu committed:
Merge pull request #88 from soborok/screenshot

added screenshot

soborok committed:
added resque auth with devise

soborok committed:
interim saving for partial use

soborok committed:
rolled back seed file to original

soborok committed:
resolved merge conflict

dalrayes committed:
Merge pull request #89 from soborok/heroku-and-partial

Heroku and partial

dalrayes committed:
Starting CSS revamp  - adding logo

soborok committed:
fixed website checkbox

pgblu committed:
new migration, fixed custom_crawler to catch encoding errors

soborok committed:
Merge pull request #90 from soborok/website-checkbox

Website checkbox

soborok committed:
Merge pull request #91 from pgblu/master

new migration, fixed custom_crawler to catch encoding errors

dalrayes committed:
first attempt at styling

Landing page has logo but needs to be centered

Home page still having issues with grid
Highlight and order of search results updated
Flag page needs to be updated to be the same

soborok committed:
Merge pull request #92 from dalrayes/master

CSS Revamp (first steps)

alcastaneda committed:
add email background job

dalrayes committed:
resized logo image - smaller

dalrayes committed:
Merge pull request #93 from alcastaneda/master

add email background job

pgblu committed:
crawler process

pgblu committed:
crawler process

pgblu committed:
Merge remote-tracking branch 'origin'

pgblu committed:
crawler digs deep

dalrayes committed:
Nav bar updated with working links

soborok committed:
resolvign merge conflict

soborok committed:
Merge pull request #94 from pgblu/master

crawler

dalrayes committed:
updated decimal places in hit score

soborok committed:
Merge pull request #96 from dalrayes/master

CSS NAV bar and update to decimal places in hit score

alcastaneda committed:
changed typo in email

soborok committed:
updated flags table to save raw_file from page

alcastaneda committed:
commented out broken target code

soborok committed:
Merge pull request #98 from soborok/update-flags

Update flags table

dalrayes committed:
CSS changes to edit query form and display of the main page

soborok committed:
Merge pull request #99 from dalrayes/master

CSS changes to edit query form and display of the main page

soborok committed:
maded adjustment in seed file, page/scraper controller for controlling heroku db

dalrayes committed:
minor changes to layout

alcastaneda committed:
fix clockwork times

alcastaneda committed:
Merge pull request #100 from dalrayes/master

minor changes to layout

soborok committed:
Merge pull request #101 from alcastaneda/master

Added email feature

dalrayes committed:
Merge pull request #102 from soborok/heroku-adjustment

Heroku adjustment

pgblu committed:
resque task

pgblu committed:
Merge branch 'resque-test'

soborok committed:
Merge pull request #103 from pgblu/master

Resque test works

dalrayes committed:
moving foundation icons

soborok committed:
adding code on search view to deploy to heroku

soborok committed:
addeding code on view files

soborok committed:
hello

soborok committed:
added gem foundation-icons-sass-rails

soborok committed:
Merge branch 'foundation-icons-to-local'

soborok committed:
Merge branch 'foundation-icons-sass'

soborok committed:
Don't diplay none flags & https support for google fonts

soborok committed:
Revert "moving foundation icons"

This reverts commit 40ebc2ff976f7e37f44a70bd8a18357605726ecf.

soborok committed:
removed debugging text

soborok committed:
added changed flag icon to unicode character

dalrayes committed:
Merge pull request #104 from soborok/flag-showing

Flag showing

soborok committed:
edited db files and clock file

soborok committed:
edited db files for heroku demo

soborok committed:
reverted clock changes

soborok committed:
cleaned for final code push to github

dalrayes committed:
Merge pull request #105 from soborok/final-cleaned-code

Final cleaned code

pgblu committed:
deleted unnecessary files, renamed task_worker

soborok committed:
Merge pull request #106 from pgblu/master

deleted unnecessary files, renamed task_worker
